go/runtime/host/composite/host.go

```diff
@@ -98,42 +113,108 @@ func (h *compositeHost) Call(ctx context.Context, body *protocol.Body) (*protoco
 	return result, nil
 }
 
-// Implements host.Runtime.
-func (h *compositeHost) UpdateCapabilityTEE() {
+// UpdateCapabilityTEE implements host.Runtime.
+func (h *Host) UpdateCapabilityTEE() {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
 	for _, comp := range h.comps {
 		comp.UpdateCapabilityTEE()
 	}
 }
 
-// Implements host.Runtime.
-func (h *compositeHost) WatchEvents() (<-chan *host.Event, pubsub.ClosableSubscription) {
-	return h.comps[component.ID_RONL].WatchEvents()
+// WatchEvents implements host.Runtime.
+func (h *Host) WatchEvents() (<-chan *host.Event, pubsub.ClosableSubscription) {
+	return h.ronl.WatchEvents()
 }
 
-// Implements host.Runtime.
-func (h *compositeHost) Start() {
+// Start implements host.Runtime.
+func (h *Host) Start() {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
 	for _, comp := range h.comps {
 		comp.Start()
 	}
+	h.running = true
 }
 
-// Implements host.Runtime.
-func (h *compositeHost) Abort(ctx context.Context, force bool) error {
+// Abort implements host.Runtime.
+func (h *Host) Abort(ctx context.Context, force bool) error {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
 	// Only RONL supports aborts.
-	return h.comps[component.ID_RONL].Abort(ctx, force)
+	return h.ronl.Abort(ctx, force)
 }
 
-// Implements host.Runtime.
-func (h *compositeHost) Stop() {
-	close(h.stopCh)
+// Stop implements host.Runtime.
+func (h *Host) Stop() {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
 
 	for _, comp := range h.comps {
 		comp.Stop()
 	}
+	h.running = false
 }
 
-// Implements host.CompositeRuntime.
-func (h *compositeHost) Component(id component.ID) (host.Runtime, bool) {
+// Component implements host.CompositeRuntime.
+func (h *Host) Component(id component.ID) (*multi.Aggregate, bool) {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
 	comp, ok := h.comps[id]
 	return comp, ok
 }
+
+// Components returns all runtime component hosts.
+func (h *Host) Components() map[component.ID]*multi.Aggregate {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
+	return maps.Clone(h.comps)
+}
+
+// HasVersion checks if the runtime component host exists for the given version.
+func (h *Host) HasVersion(id component.ID, version version.Version) bool {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+
+	comp, ok := h.comps[id]
+	if !ok {
+		return false
+	}
+	return comp.HasVersion(version)
+}
+
+// AddVersion adds a new version of the runtime component.
+func (h *Host) AddVersion(id component.ID, version version.Version, rt host.Runtime) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	h.logger.Info("adding version",
+		"id", id,
+		"version", version,
+	)
+
+	comp, ok := h.comps[id]
+	if !ok {
+		comp = multi.New(h.id)
+		if h.running {
+			comp.Start()
+		}
+		h.comps[id] = comp
+	}
+
+	if err := comp.AddVersion(version, rt); err != nil {
```

Comment:
Are old versions ever removed? I guess they are set to `nil` once stopped (in `multi.Aggregate`)?

Reply:
They are only removed if they are started. Will add cleanup in another PR.

Reply:
Currently, the ROFL components cannot be stopped and if we have a ROFL component, we always run the latest version. Therefore, it doesn't make sense to remove aggregated ROFL host from the composite host. We should add this latter when needed.

We do however need to remove obsolete RONL and ROFL versions from the aggregate. This is fixed in #6028, and can be latter changed to keep some old ROFL versions in case the scheduler wants that (or maybe the scheduler just creates a new host with an older version).